File: 1 basic/1027. 打印沙漏(20).cpp
#include<cstdio> 
#include<iostream>
#include<string>
using namespace std;


int main(){
	int total=1,count=1,N;
	char ch;
	cin>>N>>ch;
	
	//找到需要使用的符号数量 
	while(total<=N){
		count+=2;
		total+=count*2; 
	}
	total -= count*2;
	count -= 2;
	
//	cout<<"Test:count="<<count<<endl;
	
	
	//打印上三角形 
	int space=-1;
	for(int i=count; i>=1; i-=2){
		cout<<string(++space,' ')<<string(i,ch)<<endl; 
	}
	//打印下三角形 
	for(int i=3; i<=count; i+=2) {
		cout<<string(--space,' ')<<string(i,ch)<<endl;
	}
	cout<<N-total<<endl;
	
	return 0;
}

File: 1 basic/py/1027. 打印沙漏(20).py
N, ch = input().split()
N = int(N)
total,Max = 1,3
while total + Max * 2 <= N:
    total += Max * 2
    Max += 2
Max -= 2
space,i = 0, Max
while i >= 1:
    print(' ' * space + ch * i)
    space += 1
    i -= 2
i = 1
space -= 1
while i < Max:
    i += 2
    space -= 1
    print(' ' * space + ch * i)
print(N - total)